src/publish.ts

```diff
@@ -0,0 +1,323 @@
+import * as core from '@actions/core'
+import * as exec from '@actions/exec'
+import * as yaml from 'js-yaml'
+
+import * as github from '@actions/github'
+import { Plan } from './model'
+import { DefaultArtifactClient } from '@actions/artifact'
+import fs from 'fs'
+import path from 'path'
+import os from 'os'
+
+class Publish {
+  private token: string
+  private charmhubToken: string
+  private octokit
+  private workflowFile: string
+  private artifact
+  private workingDir: string
+  private resourceMapping: { [key: string]: string }
+  private workflowRunId: string
+  private identifier: string
+
+  constructor() {
+    this.token = core.getInput('github-token')
+    this.charmhubToken = core.getInput('charmhub-token')
+    this.octokit = github.getOctokit(this.token)
+    this.workflowFile = core.getInput('workflow-file')
+    this.workingDir = core.getInput('working-directory')
+    this.resourceMapping = JSON.parse(core.getInput('resource-mapping'))
+    this.workflowRunId = core.getInput('workflow-run-id')
+    this.identifier = core.getInput('identifier')
+    this.artifact = new DefaultArtifactClient()
+  }
+
+  async findWorkflowRunId(): Promise<number> {
+    if (this.workflowRunId !== '') {
+      return parseInt(this.workflowRunId)
+    }
+    const owner = github.context.repo.owner
+    const repo = github.context.repo.repo
+    const commit = await this.octokit.rest.git.getCommit({
+      owner,
+      repo,
+      commit_sha: github.context.sha
+    })
+    const tree = commit.data.tree.sha
+    const workflowResp = await this.octokit.rest.actions.getWorkflow({
+      owner,
+      repo,
+      workflow_id: this.workflowFile
+    })
+    if (workflowResp.status !== 200) {
+      throw new Error(
+        `failed to find integration workflow (${this.workflowFile}): status ${workflowResp.status}`
+      )
+    }
+    const runIter = this.octokit.paginate.iterator(
+      this.octokit.rest.actions.listWorkflowRuns,
+      {
+        owner,
+        repo,
+        workflow_id: workflowResp.data.id,
+        status: 'success',
+        created: `>${new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10)}`
```

Comment:
Typically, the publish workflow is triggered immediately following the merge of a pull request. The time limit is just a precaution to prevent the upload of outdated charm artifacts in case the publish workflow not executed for a long time after the integration test.